Let child CUDAFuture wait for parent CUDAFuture's CUDAEvents

mrshenli opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#51820 Let child CUDAFuture wait for parent CUDAFuture's CUDAEvents**

If the child cannot extract tensors from returned IValue, the
current child CUDAFuture won't wait for anything. In this case,
if the `wait()` wasn't called on the parent Future, streams are
not synchronized, and it is possible that parent Future's CUDA
ops have not been added to streams yet.

This commit adds a `markCompletedWithDataPtrs()` to `ivalue::Future`,
and RPC uses this API to pass Message tensor dataPtrs to the
`PyObject` Future when marking it as completed.

Differential Revision: [D26324068](https://our.internmc.facebook.com/intern/diff/D26324068)

mrshenli committed:
[WIP] Allow callsite to set DataPtrs for Future

[ghstack-poisoned]

mrshenli committed:
Update on "[WIP] Allow callsite to set DataPtrs for Future"




[ghstack-poisoned]

mrshenli committed:
Update on "Let child CUDAFuture wait for parent CUDAFuture's CUDAEvents"


If the child cannot extract tensors from returned IValue, the
current child CUDAFuture won't wait for anything. In this case,
if the `wait()` wasn't called on the parent Future, streams are
not synchronized, and it is possible that parent Future's CUDA
ops have not been added to streams yet.

This commit adds a `markCompletedWithDataPtrs()` to `ivalue::Future`,
and RPC uses this API to pass Message tensor dataPtrs to the
`PyObject` Future when marking it as completed.

[ghstack-poisoned]

mrshenli committed:
Update on "Let child CUDAFuture wait for parent CUDAFuture's CUDAEvents"


If the child cannot extract tensors from returned IValue, the
current child CUDAFuture won't wait for anything. In this case,
if the `wait()` wasn't called on the parent Future, streams are
not synchronized, and it is possible that parent Future's CUDA
ops have not been added to streams yet.

This commit adds a `markCompletedWithDataPtrs()` to `ivalue::Future`,
and RPC uses this API to pass Message tensor dataPtrs to the
`PyObject` Future when marking it as completed.

[ghstack-poisoned]

pritamdamania87 commented on aten/src/ATen/core/ivalue_inl.h:
```diff
@@ -355,6 +355,19 @@ struct C10_EXPORT ivalue::Future : c10::intrusive_ptr_target {
+  }
+
+  /**
+   * Explicitly mark the future as comleted with the output value and DataPtrs.
```
nit: completed

pritamdamania87 commented on aten/src/ATen/core/ivalue_inl.h:
```diff
@@ -355,6 +355,19 @@ struct C10_EXPORT ivalue::Future : c10::intrusive_ptr_target {
+   * Explicitly mark the future as comleted with the output value and DataPtrs.
+   * The data_ptrs contains storage pointers for all tensors in IValue, which
+   * will be passed to postMarkCompletedHook. Some subclass, like CUDAFuture,
+   * uses these DataPtrs to synchronize CUDA streams.
```
nit: Maybe we should mention in the docs that we should use this API only if we can't extract tensors from the IValue itself.

pritamdamania87 commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -5301,6 +5301,42 @@ def test_custom_stream_nested_multi(self):
+    @staticmethod
+    def _gpu_add_wrong_gpus(x, y):
+        if x.is_cuda and y.is_cuda:
+            return x.cpu() + y.cuda()
```
Do we need to have an RPC that errors out to test the code changes? Shouldn't a regular RPC uncover the bug and make this test fail?

mrshenli replied:
regular RPC and cuda RPC go through different code path. Regular RPC will return an `ivalue::Future`, which does not hit the subvalues problem. The CUDA RPC returns a `CUDAFuture`, and this is where we hit the error. It is true that we don't need a new static method here, but I just wanna make sure that common wrong device error on remote side can be correctly propagated to the caller, as this might be an common error when people start to use CUDA RPC.

mrshenli committed:
Update on "Let child CUDAFuture wait for parent CUDAFuture's CUDAEvents"


If the child cannot extract tensors from returned IValue, the
current child CUDAFuture won't wait for anything. In this case,
if the `wait()` wasn't called on the parent Future, streams are
not synchronized, and it is possible that parent Future's CUDA
ops have not been added to streams yet.

This commit adds a `markCompletedWithDataPtrs()` to `ivalue::Future`,
and RPC uses this API to pass Message tensor dataPtrs to the
`PyObject` Future when marking it as completed.

Differential Revision: [D26324068](https://our.internmc.facebook.com/intern/diff/D26324068)

[ghstack-poisoned]
